ArC: fix build-time and runtime resolution of beans with recursive generic types

Ladicek opened the pull request:
Fixes #17147

Includes 3 commits:

### ArC: fix assignability check during build-time resolution

The `AssignabilityCheck.isAssignableFrom()` method used to search a Jandex index for subtypes of first type to determine if it is assignable from the second type. This doesn't work well with on-demand indexing; for example, if we ask the index for all subtypes of `Comparable`, we don't get anything, because JDK types are not in the index.

This commit reverses the `isAssignableFrom()` logic: instead of searching the index for subtypes of the first type, it searches the index for supertypes of the second. That works well with on-demand indexing and doesn't change meaning.

This change alone is enough to fix build-time resolution that involves recursive generic type. For example, a `@Dependent` producer method for type `List<T>`, where `T extends Comparable<T>`, now correctly matches when resolving an injection point of type `List<String>`.

### ArC: fix runtime representation of recursive bean types

This was already attempted once, but the previous fix only worked well for simple recursive generic types. This commit fixes the runtime representation even for mutually recursive generic types.

The fix uses the following observation: in a type, multiple occurences of a type variable (as denoted by the type variable identifier) always refers to the same type. Therefore, it is not necessary to maintain a stack of type variables as they being processed. Instead, a simple string-keyed map is enough to:

1. canonicalize all occurences of a type variable;
2. canonicalize all occurences of a type variable reference;
3. at the end, patch all type variable references to point to the type variables.

The canonicalization was not performed before, but is necessarsy now (to correctly handle equality of multiple occurences of type variables), plus it also slightly improves runtime footprint of the set of bean types.

### ArC: fix bean type assignability for recursive generic types

This is adapted from [WELD-2738](https://github.com/weld/core/pull/2826).

Ladicek committed:
ArC: fix assignability check during build-time resolution

The `AssignabilityCheck.isAssignableFrom()` method used to search a Jandex
index for subtypes of first type to determine if it is assignable from
the second type. This doesn't work well with on-demand indexing; for example,
if we ask the index for all subtypes of `Comparable`, we don't get anything,
because JDK types are not in the index.

This commit reverses the `isAssignableFrom()` logic: instead of searching
the index for subtypes of the first type, it searches the index for supertypes
of the second. That works well with on-demand indexing and doesn't change
meaning.

This change alone is enough to fix build-time resolution that involves
recursive generic type. For example, a `@Dependent` producer method for
type `List<T>`, where `T extends Comparable<T>`, now correctly matches
when resolving an injection point of type `List<String>`.

Ladicek commented:
Draft because this only fixes build-time resolution and not runtime programmatic lookup. So it's only a partial fix (for now).

mkouba reviewed (commented):
Nice idea! But I think that we need to fix the runtime assignability rules as well, i.e. we should really be consistent here ;-).

manovotn commented:
@Ladicek want me to take a look into the runtime part of this?

Ladicek commented:
I actually wanted to take a look at it one of these days. But if you could take a stab at this in Weld, I guess that would be helpful as well, because a lot of the code in ArC is copied (and possibly adapted) from there.

manovotn commented:
> I actually wanted to take a look at it one of these days. But if you could take a stab at this in Weld, I guess that would be helpful as well, because a lot of the code in ArC is copied (and possibly adapted) from there.

So this is a problem in Weld as well? I might have forgotten that over the break :)
I'll take a look, thanks!

manovotn commented:
Just to keep a note on it here - I've been doing some deep dives into Weld assignability system and came with a hack that would make the scenario presented here pass.
The WIP draft can be seen at https://github.com/weld/core/pull/2826

Ladicek committed:
ArC: fix runtime representation of recursive bean types

This was already attempted once, but the previous fix only worked well
for simple recursive generic types. This commit fixes the runtime
representation even for mutually recursive generic types.

The fix uses the following observation: in a type, multiple occurences
of a type variable (as denoted by the type variable identifier) always
refers to the same type. Therefore, it is not necessary to maintain
a stack of type variables as they being processed. Instead, a simple
string-keyed map is enough to:

1. canonicalize all occurences of a type variable;
2. canonicalize all occurences of a type variable reference;
3. at the end, patch all type variable references to point to
   the type variables.

The canonicalization was not performed before, but is necessarsy now
(to correctly handle equality of multiple occurences of type variables),
plus it also slightly improves runtime footprint of the set of bean types.

Ladicek committed:
ArC: fix bean type assignability for recursive generic types

This is adapted from WELD-2738.

Ladicek commented:
This is now updated with the fix from Weld (as well as an additional fix to runtime representation of bean types in presence of recursive generic types). I'll let CI run in my fork before undrafting.

Ladicek commented:
Rebased, marked as ready, and updated the PR summary/description.

mkouba commented:
I don't think that we'd like to backport this, or?

Ladicek commented:
This is such a corner case I don't think anyone is waiting for fixes in previous releases, so I agree.

mkouba commented on independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/generic/RecursiveGenericInvalidTest.java:
```diff
@@ -0,0 +1,41 @@
+            .build();
+
+    @Test
+    public void test() {
```
It's a good practice to assert the `container.getFailure()` in the test method :-)

Ladicek replied:
Ah, I don't remember seeing that anywhere, but in hindsight, that's obvious! I'll add a fix.

Ladicek replied:
Done.

mkouba reviewed (approved):
Looks good otherwise.

Ladicek commented:
This thing keeps failing in CI here, but passes just fine in my fork: https://github.com/Ladicek/quarkus/actions/runs/3937566221 :shrug:

gsmet commented:
FWIW, this is not a test that usually fails, it's the first time I see it failing.

xstefank commented:
@Ladicek no, this shouldn't be failing.

Ladicek commented:
Gotcha, thanks! I'll continue looking.

gsmet commented:
Apparently it's all green now, cool!
